util/fpga/splice.py

```diff
@@ -0,0 +1,385 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+###########################################################################
+# Bitstream splice cache.
+#
+# Ideally, we'd rely on bazel to splice and cache bitstreams for us,
+# but:
+# - For test specification, it is more convenient to configure the
+#   test parameters with particular ROMs and OTP rather then to have
+#   explicit splice rules.  This is doubly true when one considers
+#   multiple exec_envs with different base bitstreams: its easier
+#   to specify the required ROM/OTP than to have a collection of splicing
+#   rules per exec_env.
+# - Bazel's action cache remembers the commandline that was used to create
+#   a given result.  Even if the input files between two actions and the
+#   program executed are identical, the output filenames are not, and thus
+#   the command-line is not identical.  This means if one specifies an
+#   identical splice action twice (ie: with different names), the splice
+#   will be executed twice.
+# - Vivado's `updatemen` utility is SLOW.  We can easily afford the disk
+#   space for lots of identical splices, but we cannot afford the impact
+#   on execution time.
+#
+###########################################################################
+
+import argparse
+import fcntl
+import hashlib
+import json
+import logging
+import os
+import os.path
```

Comment:
nit: do you need this since you always reference `os.path.<...>` below?

Reply:
Yes.  `os.path` is a separate module from `os` and modules don't import their submodules into the parent namespace.